pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java

```diff
@@ -27,15 +27,22 @@
 import io.netty.util.concurrent.FastThreadLocal;
 
 import java.io.IOException;
+import java.util.Objects;
 
 import org.apache.pulsar.client.api.MessageId;
 import org.apache.pulsar.common.api.proto.MessageIdData;
 import org.apache.pulsar.common.naming.TopicName;
 
 public class MessageIdImpl implements MessageId {
+    protected static final int NO_BATCH = -1;
+
     protected final long ledgerId;
     protected final long entryId;
     protected final int partitionIndex;
+    protected final int batchIndex;
+    protected final int batchSize;
+
+    protected final transient BatchMessageAcker acker;
```

Comment:
If we move all the fields of BatchMessageIdImpl to MessageIdImpl, the BatchMessageIdImpl is useless?

Reply:
If we can remove the `BatchMessageIdImpl` type completely that'd be great. It just seemed like it wouldn't be possible due to backwards compatibility concerns. I'm not sure I understand the benefit of having multiple types.